##AraGPT2

Using the link below, learn how to load araGPT2 base model.

https://huggingface.co/aubmindlab/aragpt2-base

In [2]:
!pip install arabert

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 3.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.4/126.4 kB 6.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 185.0/185.0 kB 11.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for emoji: filename=emoji-1.4.2-py3-none-any.whl size=186460 sha256=7834f16c139da8e4c42f94a8dc49db2edf2c725af689b0c87b249102664b2c5b
  Stored in directory: /root/.cache/pip/wheels/10/f0/fd/4813b1177405693e8da9cdea839f0fb64fde161380e058c827
Successfully built emoji


In [3]:
from transformers import GPT2TokenizerFast, pipeline
from transformers import GPT2LMHeadModel
from arabert.aragpt2.grover.modeling_gpt2 import GPT2LMHeadModel
from arabert.preprocess import ArabertPreprocessor

In [4]:
MODEL_NAME='aubmindlab/aragpt2-base'
arabert_prep = ArabertPreprocessor(model_name=MODEL_NAME)

text="الجزائر بلد"
text_clean = arabert_prep.preprocess(text)

model = GPT2LMHeadModel.from_pretrained(MODEL_NAME)
tokenizer = GPT2TokenizerFast.from_pretrained(MODEL_NAME)
generation_pipeline = pipeline("text-generation",model=model,tokenizer=tokenizer)


generation_pipeline(text,
    pad_token_id=tokenizer.eos_token_id,
    num_beams=10,
    max_length=200,
    top_p=0.9,
    repetition_penalty = 3.0,
    no_repeat_ngram_size = 3)[0]['generated_text']

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/843 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/553M [00:00<?, ?B/s]

Some weights of the model checkpoint at aubmindlab/aragpt2-base were not used when initializing GPT2LMHeadModel: ['ln_f.weight', 'ln_f.bias']
- This IS expected if you are initializing GPT2LMHeadModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing GPT2LMHeadModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of GPT2LMHeadModel were not initialized from the model checkpoint at aubmindlab/aragpt2-base and are newly initialized: ['emb_norm.bias', 'emb_norm.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


vocab.json:   0%|          | 0.00/1.94M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.50M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/4.52M [00:00<?, ?B/s]

'الجزائر بلد بصورةجب انهون الأسرة ، هو و أنها هي الى ان بشكل جميع هى انها القاعدةالات فيهرم أنن ولا الممثلة والمدر كافة نصف علىإ يعتبر في البيانبن القيمةأن عبر شكل النظام والاسرة ى الاسرةالإ الأصل هما قيمة بأنلت نظ عن شبه إلى نور العلامة وكل وكذلك ربينت الافرادالا� وجحيات الأم عليه الأفراد فيصل ممثلة فهي فى كل تعتبر الاصل جيهان الشكل أيضا وهو تماما 7 وتلك نورا ايضا مثل بأكملهجذا أنه علاء وبقية وتعتبر كما العائلة مريم وباقي وناالستى وهي بأكملهاث كلها وسائر او 27 تلك كثيرة نظامالت له من نتمنى 2 الملك ويتم البندهم منذ 23 الام لانعتي شكله الملاكلهالل� 11 لذلك حتى بما كيت بقية جمعاءأت فإن لة ضوءتن 2017 كبير الدولةانته مجموع والدكتورة مختلف بالإضافة بالتأكيد آ الجش اللهالع كاملالات� عموماالسلرينب بوصولالاتهاد دليل والخارج جالر أرباع ربى الل الشعب وبعض السلام ؛ منهاا وغيرها لمختلف انهما لجميع وقد ونح ء وشبه ال فيع بواسطةه بالاضافة الشوارع الاسم منه السيدة فهو وجميع وع لش وح كذلكحأ'

### Print AraGPT Model and analyze the architecture

In [5]:
print(model)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(64000, 768)
    (wpe): Embedding(1024, 768)
    (emb_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
  )
  (lm_head): Linear(in_features=768, out_features=64000, bias=False)
)


### Embedding Layers:
    The wte is for token embeddings, mapping each token to a 768-dimensional space.
    The wpe provides position embeddings, allowing the model to understand word order.

### LayerNorm: 
    The emb_norm is used to normalize the embeddings before they're passed to the transformer blocks.

### Dropout:
    Applied after embedding normalization to prevent overfitting by randomly zeroing some of the elements.

### Transformer Blocks (GPT2Block):
    There are 12 such blocks (0-11), each with its own LayerNorm, attention mechanism, and MLP.
    The attention mechanism (GPT2Attention) uses convolutions (Conv1D) for efficient calculation of query, key, and value vectors.
    Dropout is used within the attention and after the residual connections to improve generalization.

### MLP (GPT2MLP):
     A two-layer feed-forward neural network within each transformer block, with a GELU activation function.

### Output Layer (lm_head):
     A linear layer that maps the final transformer output back to the vocabulary space for token prediction.